### PR TITLE
Support cases like "next Monday or next Tuesday" in ComplexCalendar mode

### PR DESCRIPTION
**Pull request by guom08**
Javascript and Python support are not part of this PR as this feature doesn't exist on other platforms.

### CODE REVIEW

File: .NET/Microsoft.Recognizers.Text.DateTime/Extractors/BaseDateTimeAltExtractor.cs [hunk 1]
```diff
@@ -367,42 +352,118 @@ private bool IsConnectorOrWhiteSpace(int start, int end, string text)
...
+                        i = j;
                     }
                 }
             }
-            
+
+            return isValidContext;
+        }
+
+        private Dictionary<string, object> GenerateAltData(string subType, string parentText, ExtractResult contextEr = null)
```
Comment (tellarin) [inline .NET/Microsoft.Recognizers.Text.DateTime/Extractors/BaseDateTimeAltExtractor.cs]: I don't get what this means.

File: .NET/Microsoft.Recognizers.Text.DateTime/Extractors/BaseDateTimeAltExtractor.cs [hunk 2]
```diff
@@ -367,42 +352,118 @@ private bool IsConnectorOrWhiteSpace(int start, int end, string text)
...
+            if (contextEr != null)
+            {
+                data.Add(Constants.Context, contextEr);
+            }
+
             return data;
         }
 
+
+        private ExtractResult ExtractRelativePrefixContext(string text)
```
Comment (tellarin) [inline .NET/Microsoft.Recognizers.Text.DateTime/Extractors/BaseDateTimeAltExtractor.cs]: Which text is this? The whole query? Can you rename for clarity?

File: .NET/Microsoft.Recognizers.Text.DateTime/Extractors/BaseDateTimeAltExtractor.cs [hunk 3]
```diff
@@ -367,42 +352,118 @@ private bool IsConnectorOrWhiteSpace(int start, int end, string text)
...
+
+                // DatePeriod as Context: such as "next week on Tuesday or Thursday"
+                if (datePeriodErs.Count == 1)
                 {
-                    data.Add(Constants.Context, ers[0]);
-                    data.Add(Constants.SubType, Constants.SYS_DATETIME_DATE);
+                    extractResults[0].Data = GenerateAltData(extractResults[0].Type, parentText, contextEr: null);
+                    extractResults[0].Type = ExtractorName;
+
+                    for (int i = 1; i < extractResults.Count; i++)
```
Comment (tellarin) [inline .NET/Microsoft.Recognizers.Text.DateTime/Extractors/BaseDateTimeAltExtractor.cs]: Can you add some comments in this loop? I don't follow what is happening.

File: .NET/Microsoft.Recognizers.Text.DateTime/Extractors/BaseDateTimeAltExtractor.cs [hunk 4]
```diff
@@ -335,7 +300,27 @@ private bool IsConnectorOrWhiteSpace(int start, int end, string text)
                 data = ExtractDate_DateRange(former, latter);
             }
 
-            return data;
+            if (data.Count == 0)
```
Comment (tellarin) [inline .NET/Microsoft.Recognizers.Text.DateTime/Extractors/BaseDateTimeAltExtractor.cs]: Based on our offline chat, let's make sure this whole method follows the same principle. At the very lease, invert the check below so the method can be split in two separate behaviours.
Thanks.

File: .NET/Microsoft.Recognizers.Text.DateTime/Extractors/BaseDateTimeAltExtractor.cs (hunk 1 above)
Comment (guom08) [reply .NET/Microsoft.Recognizers.Text.DateTime/Extractors/BaseDateTimeAltExtractor.cs]: Changed name of this method to "CreateMetadata"

File: .NET/Microsoft.Recognizers.Text.DateTime/Extractors/BaseDateTimeAltExtractor.cs (hunk 2 above)
Comment (guom08) [reply .NET/Microsoft.Recognizers.Text.DateTime/Extractors/BaseDateTimeAltExtractor.cs]: Change this parameter name to "entityText"

**Pull request merged**